sdk/src/v2_api/builder.rs

```diff
@@ -0,0 +1,1150 @@
+    /// # Arguments
+    /// * `id` - The identifier for the resource
+    /// * `stream` - A stream to read the resource from
+    /// # Returns
+    /// * A mutable reference to the builder
+    /// # Errors
+    /// * If the resource is not valid
+    pub fn add_resource<R>(&mut self, id: &str, stream: &mut R) -> Result<&mut Self>
+    where
+        R: Read + Seek + Send,
+    {
+        if self.resources.exists(id) {
+            return Err(Error::BadParam(id.to_string())); // todo add specific error
+        }
+        let mut buf = Vec::new();
+        let _size = stream.read_to_end(&mut buf)?;
+        self.resources.add(id, buf)?;
+        Ok(self)
+    }
+
+    /// Convert the Builder into a zip formatted stream
+    /// # Arguments
+    /// * `stream` - A stream to write the zip to
+    /// # Errors
+    /// * If the zip cannot be written
+    pub fn zip(&mut self, stream: impl Write + Seek) -> Result<()> {
+        drop(
+            // this drop seems to be required to force a flush before reading back.
+            {
+                let mut zip = ZipWriter::new(stream);
+                let options =
+                    FileOptions::default().compression_method(zip::CompressionMethod::Stored);
+                zip.start_file("manifest.json", options)
+                    .map_err(|e| Error::OtherError(Box::new(e)))?;
+                zip.write_all(&serde_json::to_vec(self)?)?;
+                // add a folder to the zip file
+                zip.start_file("resources/", options)
+                    .map_err(|e| Error::OtherError(Box::new(e)))?;
+                for (id, data) in self.resources.resources() {
+                    zip.start_file(format!("resources/{}", id), options)
+                        .map_err(|e| Error::OtherError(Box::new(e)))?;
+                    zip.write_all(data)?;
+                }
+                for (index, ingredient) in self.definition.ingredients.iter().enumerate() {
+                    zip.start_file(format!("ingredients/{}/", index), options)
+                        .map_err(|e| Error::OtherError(Box::new(e)))?;
+                    for (id, data) in ingredient.resources().resources() {
+                        //println!("adding ingredient {}/{}", index, id);
+                        zip.start_file(format!("ingredients/{}/{}", index, id), options)
+                            .map_err(|e| Error::OtherError(Box::new(e)))?;
+                        zip.write_all(data)?;
+                    }
+                }
+                zip.finish()
+            }
+            .map_err(|e| Error::OtherError(Box::new(e)))?,
+        );
+        Ok(())
+    }
+
+    //#[allow(clippy::unwrap_used)]
+    /// Unzip a stream into a Builder
+    /// # Arguments
+    /// * `stream` - A stream to read the zip from
+    /// # Returns
+    /// * A new Builder
+    /// # Errors
+    /// * If the zip cannot be read
+    pub fn unzip(stream: impl Read + Seek) -> Result<Self> {
```

Comment:
Do you think these functions would be clearer/more consistent if they were called something like `to_archive` and `from_archive`?

Reply:
yes I probably should not use the term "zip" for these. I'll change that.